### PR TITLE
[JENKINS-68126] Remove event streaming causing performance problems at scale

### PR DESCRIPTION
**Pull request by sbeaulie**
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

See ticket [JENKINS-68126](https://issues.jenkins.io/browse/JENKINS-68126) for background info.

Pre-requisite
Websocket needs disabled

Before the change when a lot of load (600+ queued jobs) wanted to start, a few would make it but then the k8s plugin would stop processing `launch`. Threads would show waiting inside the watcher. Adding logs to w1 and w2 also confirmed that something would wait indefinitely under load. The result were a few nodes running tasks but the majority showing as idle / suspended (isAcceptingTask equals to false).

With the change the load graphs look much better.
Screenshot 2022-04-25 at 11 42 37

Thanks to the team and specially @suckatrash @jonathannewman

### CODE REVIEW

**Comment by sbeaulie** [comment]
The original test we did was to separate the try-with-resources block for the watchers and add some logging - see initial commit commented out (for historical view in the commit)

https://github.com/jenkinsci/kubernetes-plugin/pull/1167/commits/4a298226f447bf69571a01631249cd29065ade7a#diff-8e43be7dacff0fa974a7614fda2f4103bf4b820be52d747d9498e6f7efbd688fR179-R182

These w1 and w2 object would not return, and so we would never see the log messages, past the first maybe 20-90 nodes when under load (200-600 queued jobs)

**Comment by itamarbareket** [comment]
This is gold. Performance is much much better. 
image

File: pom.xml [hunk 1]
```diff
@@ -39,7 +39,7 @@
   </developers>
 
   <properties>
-    <changelist>999999-SNAPSHOT</changelist>
+    <changelist>68126-SNAPSHOT-NOWATCHER</changelist>
```
Comment (jglick) [inline pom.xml:42]: Revert this. If tests pass in the PR, and it is up to date with `master`, you will get an incremental deployment automatically (link displayed among status checks).

Comment (jglick) [reply pom.xml:42]: (Currently there are four failures in https://ci.jenkins.io/job/Plugins/job/kubernetes-plugin/job/PR-1167/6/testReport/org.csanchez.jenkins.plugins.kubernetes.pipeline/KubernetesPipelineTest/.)

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java [hunk 2]
```diff
@@ -166,12 +165,13 @@ else if (httpCode == 409 && e.getMessage().contains("Operation cannot be fulfill
             ObjectMeta podMetadata = pod.getMetadata();
             template.getWorkspaceVolume().createVolume(client, podMetadata);
             template.getVolumes().forEach(volume -> volume.createVolume(client, podMetadata));
-            watcher = new AllContainersRunningPodWatcher(client, pod);
-            try (Watch w1 = client.pods().inNamespace(namespace).withName(podName).watch(watcher);
-                 Watch w2 = eventWatch(client, podName, namespace, runListener)) {
-                assert watcher != null; // assigned 3 lines above
-                watcher.await(template.getSlaveConnectTimeout(), TimeUnit.SECONDS);
+
+            // Removed watcher because of load issues making the fabric8 library wait indefinitely https://issues.jenkins.io/browse/JENKINS-68126
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:169]:
```suggestion
```

Comment meaningless in the future state—only meaningful for the diff (PR comment).

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java [hunk 3]
```diff
@@ -256,17 +256,7 @@ else if (httpCode == 409 && e.getMessage().contains("Operation cannot be fulfill
             Metrics.metricRegistry().counter(MetricNames.PODS_LAUNCHED).inc();
         } catch (Throwable ex) {
             setProblem(ex);
-            if (ex instanceof AllContainersRunningPodWatcher.PodNotRunningException) {
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:259]: (From #1061, not now relevant AFAICT.)

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/AllContainersRunningPodWatcher.java [hunk 4]
```diff
@@ -1,166 +0,0 @@
...
-            PodNotRunningException x = new PodNotRunningException(String.format("Pod has terminated containers: %s/%s (%s)",
-                    this.pod.getMetadata().getNamespace(),
-                    this.pod.getMetadata().getName(),
-                    terminatedContainers.stream()
-                            .map(ContainerStatus::getName)
-                            .collect(joining(", ")
-                            )));
-            String logs = PodUtils.logLastLines(this.pod, client);
-            if (logs != null) {
-                x.addSuppressed(new ContainerLogs(logs));
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/AllContainersRunningPodWatcher.java:135]: Makes `ContainerLogs` dead code AFAICT?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java [hunk 5]
```diff
@@ -166,12 +165,13 @@ else if (httpCode == 409 && e.getMessage().contains("Operation cannot be fulfill
...
-            try (Watch w1 = client.pods().inNamespace(namespace).withName(podName).watch(watcher);
-                 Watch w2 = eventWatch(client, podName, namespace, runListener)) {
-                assert watcher != null; // assigned 3 lines above
-                watcher.await(template.getSlaveConnectTimeout(), TimeUnit.SECONDS);
+
+            // Removed watcher because of load issues making the fabric8 library wait indefinitely https://issues.jenkins.io/browse/JENKINS-68126
+            Pod checkPod = client.pods().inNamespace(namespace).withName(podName).waitUntilReady(template.getSlaveConnectTimeout(), TimeUnit.SECONDS);
+            if (!Readiness.isPodReady(checkPod)) {
+                throw new KubernetesClientTimeoutException(pod, template.getSlaveConnectTimeout(), TimeUnit.SECONDS);
             }
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:173]: I'm not sure this is even necessary, the exception should be raised already by `waitUntilReady` if it reaches the timeout.

**Review by Vlatombe — COMMENTED** [review]
Looks reasonable as long as unit tests keep passing.
Event streaming was mostly useful for diagnostic purpose, for people without direct access to the underlying Kubernetes cluster.
However most of the error conditions are now handled via `Reaper.Listener` implementations , so I believe we could live without it.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java (hunk 5 above)
Comment (Vlatombe) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesLauncher.java:173]: I'm not sure...

File: pom.xml (hunk 1 above)
Comment (sbeaulie) [reply pom.xml]: sorry I was on PTO, done

**Comment by Vlatombe** [comment]
@sbeaulie Can you fix the unit tests please?

**Pull request merged**